Sources/MapboxDirections/DirectionsResult.swift

```diff
@@ -0,0 +1,182 @@
+import Foundation
+import struct Polyline.Polyline
+import CoreLocation
+import struct Turf.LineString
+
+/**
+ A `DirectionsResult` represents a result returned from either the Mapbox Directions service.
+ 
+ You do not create instances of this class directly. Instead, you receive `Route` or `Match` objects when you request directions using the `Directions.calculate(_:completionHandler:)` or `Directions.calculateRoutes(matching:completionHandler:)` method.
+ */
+open class DirectionsResult: Codable {
+    private enum CodingKeys: String, CodingKey {
+        case shape = "geometry"
+        case legs
+        case distance
+        case expectedTravelTime = "duration"
+        case directionsOptions
+        case accessToken
+        case apiEndpoint
+        case routeIdentifier
+        case speechLocale = "voiceLocale"
+    }
+    
+    // MARK: Creating a Directions Result
+    
+    public required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        legs = try container.decode([RouteLeg].self, forKey: .legs)
+        distance = try container.decode(CLLocationDistance.self, forKey: .distance)
+        expectedTravelTime = try container.decode(TimeInterval.self, forKey: .expectedTravelTime)
+        
+        _directionsOptions = decoder.userInfo[.options] as! DirectionsOptions
```

Comment:
This PR doesn’t completely fix #228 because it still contains some force-unwrapping of user info dictionary entries. This is problematic because decoding from JSON is now part of the public API for all types. Any user info dictionary entries we rely on for decoding need to be documented as preconditions, and the absence of them should throw an error instead of crashing. Both the error-throwing and documentation are addressed in #388, but we should try to bring those changes over to this PR because they’re more relevant to the decoding changes that are taking place here.

Reply:
c5e733ddd67e80a18316b133823982298f7dcdb0 brings over the error-throwing and documentation from #388.